src/library_pthread.js

```diff
@@ -678,14 +687,14 @@ var LibraryPThread = {
     msg.moduleCanvasId = threadParams.moduleCanvasId;
     msg.offscreenCanvases = threadParams.offscreenCanvases;
 #endif
-    // Ask the worker to start executing its pthread entry point function.
 #if ENVIRONMENT_MAY_BE_NODE
     if (ENVIRONMENT_IS_NODE) {
-      // Mark worker as strongly referenced once we start executing a pthread,
-      // so that Node.js doesn't exit while the pthread is running.
-      worker.ref();
+      // Mark worker as weakly referenced once we start executing a pthread,
```

Comment:
What I don't understand about this comment "Mark worker as weakly referenced once we start executing a pthread" is that workers are already weakly referenced when they are in the pool, so why does anything need to change once we start executing a pthread.   They are weakly referenced both before and after they start executing a thread, no?

Reply:
IIUC, only idle/unused workers are weakly referenced within the pool (i.e. when `-sPTHREAD_POOL_SIZE=x` is specified).
https://github.com/emscripten-core/emscripten/blob/22893f5aae7f3d8aa3f13997bf3921d189879e5d/src/library_pthread.js#L292-L295
https://github.com/emscripten-core/emscripten/blob/22893f5aae7f3d8aa3f13997bf3921d189879e5d/src/library_pthread.js#L216-L219

But when `-sPTHREAD_POOL_SIZE=x` is omitted (which should be safe on Node.js after commit 73eaf8188f8f85602f37a4e8e8e347727fed968a) or when an application spawns more threads than specified within that setting (i.e. thread pool exhaustion), workers are strongly referenced by default, so this `.unref()` is still necessary for those cases.

Reply:
Can we instread fix the `workers are strongly referenced by default` thing?  

IRC the intent prior this change is that they should be weakly referenced until they are running a pthread, and then strongly referenced one a pthread is running.

After this change they should **never** be strongly referenced so should be `.unref()` as soon as we create them?

Reply:
I tried to simplify this with commit 059448ad013ab6ed5bc5efd8dd3ad6bb15c63429.

> After this change they should **never** be strongly referenced so should be `.unref()` as soon as we create them?

AFAIK it's only safe to `.unref()` the worker during the `'loaded'` event or just before executing the pthread entry point function (i.e. only after the `.on('message')` listeners are attached), it will not work if you `.unref()` too early. See for example this program:
<details>
  <summary>Details</summary>
  
```js
const { Worker, isMainThread, parentPort } = require('worker_threads');

if (isMainThread) {
  const worker = new Worker(__filename);
  worker.on('error', (err) => { throw err; });
  worker.on('exit', () => console.log('exit'));
  worker.on('message', (msg) => console.log(msg))
  worker.postMessage('From main thread!')
  /* setTimeout(() => */worker.unref()/* , 100); */
} else {
  parentPort.on('message', (msg) => console.log(msg));
  parentPort.postMessage('From worker!');
}
```
</details>

As a possible enhancement, perhaps we could make `PROXY_TO_PTHREAD` no-op on Node.js (i.e. just call the original main)? This would be safe after commit 73eaf8188f8f85602f37a4e8e8e347727fed968a, and would allow further simplification.